File: Python/graphs/Floyd_Warshall_Algorithm.py
# Importing sys module to initalize the maxsize
import sys
''' Floyd Warshall Algorithm '''
''' Problem Statement :The problem is to find shortest distances
                      between every pair of vertices in a given
                      edge weighted directed Graph. The Graph is
                      represented as Adjancency Matrix, and the
                      Matrix denotes the weight of the edegs
                      (if it exists) else INF (1e7).
'''
# Assigning max value a variable can take
INT_MAX = sys.maxsize - 1
# Initialize Infinity
INF = 9999999
# Enter the size of the matrix
print("Enter the sides of the Matrix : ")
print("Enter the Row and the Column")
row = int(input("Row:"))
col = int(input("Column:"))
# Taking input of list
print("Enter the integers:")
graph = list(map(str, input().rstrip().rsplit()))
# Convert the given list into integers
for i in range(row*col):
    if(graph[i] == "INF"):
        graph[i] = INF
    else:
        graph[i] = int(graph[i])
# Constructing a Matrix.
graph = [graph[i:i + col] for i in range(0, len(graph), col)]
# Floyd Warshall


def FloydWarshall(graph, row, col):
    table = list(map(lambda i: list(map(lambda j: j, i)), graph))
    # Add the vertices
    for k in range(row):
        for i in range(col):
            for j in range(row):
                table[i][j] = min(table[i][j], table[i][k] + table[k][j])
    return(table)
# Print List Function


def printlist(table):
    for i in table:
        print(*i)
FW = FloydWarshall(graph, row, col)
print("The shortest paths are".format(printlist(FW)))
'''
Sample input:
    Enter the sides of the Matrix :
    Enter the Row and the Column
    Row:4
    Column:4
    Enter the integers:
    0 3 INF 5 2 0 INF 4 INF 1 0 INF INF INF 2 0
Sample Output:
    The shortest paths are
    0 3 7 5
    2 0 6 4
    3 1 0 5
    5 3 2 0
'''

File: Java/graphs/Floyd_Warshall_Algorithm.java
/**

The Floyd Warshall Algorithm is used to find minimum distances between
every pair of vertices in a given edge weighted directed Graph.
The Floyd-Warshall algorithm is an example of dynamic programming. 
It breaks the problem down into smaller subproblems, then combines 
the answers to those subproblems to solve the big, initial problem. 

*/

import java.util.Scanner;

public class Floyd_Warshall_Algorithm {

	public static void display(int v, int[][] answerMatrix) {

		System.out.println("Matrix displaying minimum costs between the pairs of vertices.");

		// display final matrix after all the updations of shortest cost path from
		// vertex i to j
		// value at i,j represents minimum cost from vertex i to vertex j.
		for (int i = 0; i < v; i++) {
			for (int j = 0; j < v; j++) {
				System.out.print(answerMatrix[i][j] + " ");
			}
			System.out.println();
		}
	}

	public static void floydWarshallAlgorithm(int v, int[][] graph) {

		// answer matrix where we will be storing minimum cost between two pair of
		// vertices.
		int[][] answerMatrix = new int[v][v];

		// just copy down values of graph into ans matrix.
		for (int i = 0; i < v; i++) {
			for (int j = 0; j < v; j++) {
				answerMatrix[i][j] = graph[i][j];
			}
		}

		// update the matrix by keeping each vertex as a main rout part once.
		for (int k = 0; k < v; k++) {
			// k represents that we are keeping kth vertex constant.
			// means that path from i to j must pass from k.

			for (int i = 0; i < v; i++) {
				for (int j = 0; j < v; j++) {

					int oc = answerMatrix[i][j];
					int nc = answerMatrix[i][k] + answerMatrix[k][j];

					if (nc < oc) {
						answerMatrix[i][j] = nc;
					}
				}
			}
		}

		display(v, answerMatrix);
	}

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		System.out.println("Enter no. of vertices in the Graph.");
		int v = sc.nextInt();

		int[][] graph = new int[v][v];

		System.out.println("Enter costs of edges in the Graph.");

		for (int i = 0; i < v; i++) {
			for (int j = 0; j < v; j++) {
				graph[i][j] = sc.nextInt();
			}
		}

		floydWarshallAlgorithm(v, graph);
	}
}

/**
 
Time Complexity : O(V^3)
Space Complexity : O(V^2)

Sample Test case 1 :

4
0 3 10000 7
8 0 2 10000
5 10000 0 1
2 10000 10000 0

Output :

Matrix displaying minimum costs between the pairs of vertices.
0 3 5 6 
5 0 2 3 
3 6 0 1 
2 5 7 0 

*/
